File: source/sdk/go/expressions/sexp.go
package expressions

import (
	"fmt"
	"strings"
)

// The AST as an S-expression.
//
// Parse scenarios have to assert on tree *shape* — precedence and
// associativity are the bugs they exist to catch — and a nested YAML literal of
// the whole node graph is unreadable enough that nobody would notice it
// asserting the wrong thing. `(- (- 1 2) 3)` says "folds left" at a glance.
//
// This is a contract, not a convenience: packages/expressions/src/sexp.ts must
// print the same string for the same source, and the shared scenarios are what
// check that it does.

// SexpOptions controls the rendering. Offsets appends `@offset` to every node;
// only the offset scenarios ask for it.
type SexpOptions struct {
	Offsets bool
}

// TemplateToSexp renders a whole Template.
func TemplateToSexp(template *Template, options SexpOptions) string {
	if len(template.Segments) == 0 {
		return "(template)"
	}
	parts := make([]string, 0, len(template.Segments))
	for _, segment := range template.Segments {
		parts = append(parts, segmentToSexp(segment, options))
	}
	return "(template " + strings.Join(parts, " ") + ")"
}

func segmentToSexp(segment Segment, options SexpOptions) string {
	switch s := segment.(type) {
	case *Text:
		return "(text " + quoteSexp(s.Value) + ")"
	case *Hole:
		return "(hole" + mark(s.At, options) + " " + ToSexp(s.Expr, options) + ")"
	}
	return "(unknown)"
}

// ToSexp renders one Expression.
func ToSexp(node Expression, options SexpOptions) string {
	at := mark(node.Offset(), options)
	nested := func(child Expression) string { return ToSexp(child, options) }

	switch n := node.(type) {
	case *Name:
		return n.Name + at
	case *Literal:
		return literalToSexp(n) + at
	case *Member:
		return "(." + at + " " + nested(n.Object) + " " + n.Name + ")"
	case *Index:
		return "([]" + at + " " + nested(n.Object) + " " + nested(n.Index) + ")"
	case *Project:
		return "(project" + at + " " + nested(n.Object) + ")"
	case *Call:
		parts := []string{nested(n.Object)}
		for _, arg := range n.Args {
			parts = append(parts, nested(arg))
		}
		return "(call" + at + " " + strings.Join(parts, " ") + ")"
	case *Unary:
		op := n.Op
		if op == "-" {
			op = "neg"
		}
		return "(" + op + at + " " + nested(n.Operand) + ")"
	case *Binary:
		return "(" + n.Op + at + " " + nested(n.Left) + " " + nested(n.Right) + ")"
	case *Ternary:
		return "(?:" + at + " " + nested(n.Cond) + " " + nested(n.WhenTrue) + " " + nested(n.WhenFalse) + ")"
	}
	return "(unknown)"
}

func mark(at int, options SexpOptions) string {
	if !options.Offsets {
		return ""
	}
	return fmt.Sprintf("@%d", at)
}

func literalToSexp(literal *Literal) string {
	switch literal.Kind {
	case LiteralText:
		return quoteSexp(literal.Value.(string))
	case LiteralNumber:
		return NumberToText(literal.Value.(float64))
	case LiteralBoolean:
		if literal.Value.(bool) {
			return "true"
		}
		return "false"
	}
	return "null"
}

// quoteSexp is deliberately minimal, and identical in TypeScript: backslash,
// quote and the three whitespace escapes, so a scenario expectation always fits
// on one line.
func quoteSexp(value string) string {
	replacer := strings.NewReplacer(
		`\`, `\\`,
		`"`, `\"`,
		"\n", `\n`,
		"\r", `\r`,
		"\t", `\t`,
	)
	return `"` + replacer.Replace(value) + `"`
}

File: source/packages/expressions/src/sexp.ts
/**
 * The AST as an S-expression.
 *
 * Parse scenarios have to assert on tree *shape* — precedence and
 * associativity are the bugs they exist to catch — and a nested YAML literal of
 * the whole node graph is unreadable enough that nobody would notice it
 * asserting the wrong thing. `(- (- 1 2) 3)` says "folds left" at a glance.
 *
 * This is a contract, not a convenience: `sdk/go/expressions/sexp.go` must
 * print the same string for the same source, and the shared scenarios are what
 * check that it does.
 */
import type { Expression, Segment, TemplateNode } from './ast.js'
import { numberToText } from './value.js'

export interface SexpOptions {
  /** Append `@offset` to every node. Only the offset scenarios ask for this. */
  readonly offsets?: boolean
}

export function templateToSexp(template: TemplateNode, options: SexpOptions = {}): string {
  const parts = template.segments.map((segment) => segmentToSexp(segment, options))
  return `(template ${parts.join(' ')})`.replace('(template )', '(template)')
}

function segmentToSexp(segment: Segment, options: SexpOptions): string {
  if (segment.kind === 'Text') return `(text ${quote(segment.value)})`
  return `(hole${mark(segment.at, options)} ${toSexp(segment.expr, options)})`
}

export function toSexp(node: Expression, options: SexpOptions = {}): string {
  const at = mark(node.at, options)
  const nested = (child: Expression) => toSexp(child, options)

  switch (node.kind) {
    case 'Name':
      return `${node.name}${at}`
    case 'Literal':
      return `${literal(node.type, node.value)}${at}`
    case 'Member':
      return `(.${at} ${nested(node.object)} ${node.name})`
    case 'Index':
      return `([]${at} ${nested(node.object)} ${nested(node.index)})`
    case 'Project':
      return `(project${at} ${nested(node.object)})`
    case 'Call':
      return `(call${at} ${[node.object, ...node.args].map(nested).join(' ')})`
    case 'Unary':
      return `(${node.op === '-' ? 'neg' : node.op}${at} ${nested(node.operand)})`
    case 'Binary':
      return `(${node.op}${at} ${nested(node.left)} ${nested(node.right)})`
    case 'Ternary':
      return `(?:${at} ${nested(node.cond)} ${nested(node.then)} ${nested(node.otherwise)})`
  }
}

const mark = (at: number, options: SexpOptions): string => (options.offsets ? `@${at}` : '')

function literal(type: string, value: string | number | boolean | null): string {
  if (type === 'text') return quote(String(value))
  if (type === 'number') return numberToText(value as number)
  return String(value)
}

/**
 * Deliberately minimal, and identical in Go: backslash, quote and the three
 * whitespace escapes, so a scenario expectation always fits on one line.
 */
function quote(value: string): string {
  const escaped = value
    .replace(/\\/g, '\\\\')
    .replace(/"/g, '\\"')
    .replace(/\n/g, '\\n')
    .replace(/\r/g, '\\r')
    .replace(/\t/g, '\\t')
  return `"${escaped}"`
}
